Unable to build on Solaris 11

using profile: buildconf/core.ini
detected include path: ['/opt/csw/lib/gcc/i386-pc-solaris2.10/4.9.2/include', '/usr/local/include', '/opt/csw/include', '/opt/csw/lib/gcc/i386-pc-solaris2.10/4.9.2/include-fixed', '/usr/include']
detected CPU cores: 16
configured CFLAGS: -O2 -I. -Wall -Werror -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -fno-strict-aliasing -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -DUWSGI_HAS_IFADDRS -DUWSGI_ZLIB -DUWSGI_LOCK_USE_MUTEX -DUWSGI_EVENT_USE_PORT -DUWSGI_EVENT_TIMER_USE_PORT -DUWSGI_EVENT_FILEMONITOR_USE_PORT -I/usr/include/pcre -DUWSGI_PCRE -DUWSGI_ROUTING -DUWSGI_UUID -DUWSGI_VERSION="\"2.1\"" -DUWSGI_VERSION_BASE="2" -DUWSGI_VERSION_MAJOR="1" -DUWSGI_VERSION_MINOR="0" -DUWSGI_VERSION_REVISION="0" -DUWSGI_VERSION_CUSTOM="\"\"" -DUWSGI_YAML -DUWSGI_SSL -I/usr/include/libxml2 -DUWSGI_XML -DUWSGI_XML_LIBXML2 -DUWSGI_PLUGIN_DIR="\".\""
*** uWSGI compiling server core ***
[thread 2][/opt/csw/bin/gcc-4.9] core/utils.o
[thread 0][/opt/csw/bin/gcc-4.9] core/protocol.o
[thread 1][/opt/csw/bin/gcc-4.9] core/socket.o
[thread 4][/opt/csw/bin/gcc-4.9] core/logging.o
[thread 5][/opt/csw/bin/gcc-4.9] core/master.o
[thread 6][/opt/csw/bin/gcc-4.9] core/master_utils.o
[thread 8][/opt/csw/bin/gcc-4.9] core/emperor.o
[thread 9][/opt/csw/bin/gcc-4.9] core/mule.o
[thread 7][/opt/csw/bin/gcc-4.9] core/notify.o
[thread 11][/opt/csw/bin/gcc-4.9] core/subscription.o
[thread 10][/opt/csw/bin/gcc-4.9] core/stats.o
[thread 12][/opt/csw/bin/gcc-4.9] core/sendfile.o
[thread 13][/opt/csw/bin/gcc-4.9] core/async.o
[thread 14][/opt/csw/bin/gcc-4.9] core/master_checks.o
[thread 15][/opt/csw/bin/gcc-4.9] core/fifo.o
[thread 3][/opt/csw/bin/gcc-4.9] core/offload.o
[thread 12][/opt/csw/bin/gcc-4.9] core/io.o
[thread 7][/opt/csw/bin/gcc-4.9] core/static.o
[thread 15][/opt/csw/bin/gcc-4.9] core/websockets.o
[thread 13][/opt/csw/bin/gcc-4.9] core/spooler.o
[thread 14][/opt/csw/bin/gcc-4.9] core/snmp.o
[thread 9][/opt/csw/bin/gcc-4.9] core/exceptions.o
[thread 3][/opt/csw/bin/gcc-4.9] core/config.o
[thread 10][/opt/csw/bin/gcc-4.9] core/setup_utils.o
[thread 0][/opt/csw/bin/gcc-4.9] core/clock.o
[thread 5][/opt/csw/bin/gcc-4.9] core/init.o
[thread 15][/opt/csw/bin/gcc-4.9] core/buffer.o
[thread 7][/opt/csw/bin/gcc-4.9] core/reader.o
[thread 14][/opt/csw/bin/gcc-4.9] core/writer.o
[thread 10][/opt/csw/bin/gcc-4.9] core/alarm.o
[thread 11][/opt/csw/bin/gcc-4.9] core/cron.o
[thread 9][/opt/csw/bin/gcc-4.9] core/hooks.o
[thread 0][/opt/csw/bin/gcc-4.9] core/plugins.o
[thread 13][/opt/csw/bin/gcc-4.9] core/lock.o
[thread 6][/opt/csw/bin/gcc-4.9] core/cache.o
[thread 12][/opt/csw/bin/gcc-4.9] core/daemons.o
core/hooks.c:392:12: error: conflicting types for ‘sethostname’
 extern int sethostname(char *, int);
            ^
In file included from ./uwsgi.h:229:0,
                 from core/hooks.c:1:
/usr/include/unistd.h:473:12: note: previous declaration of ‘sethostname’ was here
 extern int sethostname(const char *, size_t);
            ^
[thread 5][/opt/csw/bin/gcc-4.9] core/errors.o
[thread 4][/opt/csw/bin/gcc-4.9] core/hash.o

--- a/core/hooks.c
+++ b/core/hooks.c
@@ -396,9 +396,10 @@ static int uwsgi_hook_chown2(char *arg) {
 }
 
 
-#ifdef __sun__
+#if defined(UWSGI_SUNOS_EXTERN_SETHOSTNAME)
 extern int sethostname(char *, int);
 #endif
+
 static int uwsgi_hook_hostname(char *arg) {
 #ifdef __CYGWIN__
 	return -1;

--- a/uwsgiconfig.py
+++ b/uwsgiconfig.py
@@ -810,7 +810,12 @@ class uConf(object):
             self.libs.append('-lsendfile')
             self.libs.append('-lrt')
             self.gcc_list.append('lib/sun_fixes')
-            self.ldflags.append('-L/lib')
+            sunos_major = int(uwsgi_os_k.split('.')[0])
+            sunos_minor = int(uwsgi_os_k.split('.')[1])
+            # solaris < 11 does not have sethostname declared in unistd
+            if not (sunos_major == 5 and sunos_minor > 10):
+                self.cflags.append('-DUWSGI_SUNOS_EXTERN_SETHOSTNAME')
+                self.ldflags.append('-L/lib')
             if not uwsgi_os_v.startswith('Nexenta'):
                 self.libs.remove('-rdynamic')
 

